### PR TITLE
Add some board orientation options

### PR DESCRIPTION
**Pull request by jaxxzer**
3 new options:
 - PITCH_45
 - ROLL_90_PITCH_45
 - CUSTOM

When CUSTOM is selected, the board orientation is defined by the new AHRS_ROT_R,P,Y parameters. If custom is not selected these r,p,y parameters do nothing.

There are some comments regarding a 'complete set' of rotations at the beginning of the Rotation enum, but I can't find anywhere where this property was required/made useful. There is also a 'custom' rotation added that broke this property before I did here.

### CODE REVIEW

**Comment by rmackay9** [comment]
LGTM.

File: libraries/AP_AHRS/AP_AHRS.cpp [hunk 1]
```diff
@@ -185,6 +212,42 @@ void AP_AHRS::add_trim(float roll_in_radians, float pitch_in_radians, bool save_
...
+    if (orientation != ROTATION_CUSTOM) {
+        _ins.set_board_orientation(orientation);
+        if (_compass != nullptr) {
+            _compass->set_board_orientation(orientation);
+        }
+    } else {
+        float xrad = DEG_TO_RAD * _rot_roll;
+        float yrad = DEG_TO_RAD * _rot_pitch;
+        float zrad = DEG_TO_RAD * _rot_yaw;
+        Matrix3f rotationx = Matrix3f(
```
Comment (tridge) [inline libraries/AP_AHRS/AP_AHRS.cpp]: is there some reason we can't use the existing euler to matrix functions?

File: libraries/AP_AHRS/AP_AHRS.cpp [hunk 2]
```diff
@@ -128,6 +128,33 @@ const AP_Param::GroupInfo AP_AHRS::var_info[] = {
     AP_GROUPINFO("EKF_TYPE",  14, AP_AHRS, _ekf_type, 2),
 #endif
 
+    // @Param: ROT_ROLL
```
Comment (tridge) [inline libraries/AP_AHRS/AP_AHRS.cpp]: I think a name AHRS_CUSTOM_ROLL would be clearer (most users look only at the var name, not the docs)

File: libraries/AP_AHRS/AP_AHRS.h [hunk 3]
```diff
@@ -596,6 +591,11 @@ class AP_AHRS
     AP_Int8 _gps_minsats;
     AP_Int8 _gps_delay;
     AP_Int8 _ekf_type;
+    AP_Float _rot_roll;
```
Comment (tridge) [inline libraries/AP_AHRS/AP_AHRS.h]: better to rename these to _custom_roll etc

File: libraries/AP_Compass/AP_Compass.h [hunk 4]
```diff
@@ -204,8 +204,11 @@ friend class AP_Compass_Backend;
     float get_declination() const;
 
     // set overall board orientation
-    void set_board_orientation(enum Rotation orientation) {
+    void set_board_orientation(enum Rotation orientation, Matrix3f* custom_rotation = nullptr) {
         _board_orientation = orientation;
+        if (custom_rotation) {
```
Comment (tridge) [inline libraries/AP_Compass/AP_Compass.h]: this should be unconditional, so if the user changes away from a custom rotation to a non-custom the variable gets set back to nullptr

File: libraries/AP_InertialSensor/AP_InertialSensor.h [hunk 5]
```diff
@@ -187,8 +187,11 @@ class AP_InertialSensor : AP_AccelCal_Client
     static const struct AP_Param::GroupInfo var_info[];
 
     // set overall board orientation
-    void set_board_orientation(enum Rotation orientation) {
+    void set_board_orientation(enum Rotation orientation, Matrix3f* custom_rotation = nullptr) {
         _board_orientation = orientation;
+        if (custom_rotation) {
```
Comment (tridge) [inline libraries/AP_InertialSensor/AP_InertialSensor.h]: same here

**Comment by tridge** [comment]
also, I agree with Randy on CUSTOM being a much later enum, say 100

File: libraries/AP_AHRS/AP_AHRS.cpp (hunk 1 above)
Comment (jaxxzer) [reply libraries/AP_AHRS/AP_AHRS.cpp]: where can I find these?

File: libraries/AP_AHRS/AP_AHRS.cpp (hunk 2 above)
Comment (jaxxzer) [reply libraries/AP_AHRS/AP_AHRS.cpp]: done

File: libraries/AP_AHRS/AP_AHRS.h (hunk 3 above)
Comment (jaxxzer) [reply libraries/AP_AHRS/AP_AHRS.h]: done

File: libraries/AP_Compass/AP_Compass.h (hunk 4 above)
Comment (jaxxzer) [reply libraries/AP_Compass/AP_Compass.h]: done, good point

File: libraries/AP_InertialSensor/AP_InertialSensor.h (hunk 5 above)
Comment (jaxxzer) [reply libraries/AP_InertialSensor/AP_InertialSensor.h]: done

**Comment by jaxxzer** [comment]
> also, I agree with Randy on CUSTOM being a much later enum, say 100

done

**Comment by jaxxzer** [comment]
All done except for my diy matrix math. An example usage of the current methods for this math will be helpful for me.

File: libraries/AP_AHRS/AP_AHRS.cpp [hunk 6]
```diff
@@ -94,7 +94,7 @@ const AP_Param::GroupInfo AP_AHRS::var_info[] = {
...
+    // @Values: 0:None,1:Yaw45,2:Yaw90,3:Yaw135,4:Yaw180,5:Yaw225,6:Yaw270,7:Yaw315,8:Roll180,9:Roll180Yaw45,10:Roll180Yaw90,11:Roll180Yaw135,12:Pitch180,13:Roll180Yaw225,14:Roll180Yaw270,15:Roll180Yaw315,16:Roll90,17:Roll90Yaw45,18:Roll90Yaw90,19:Roll90Yaw135,20:Roll270,21:Roll270Yaw45,22:Roll270Yaw90,23:Roll270Yaw135,24:Pitch90,25:Pitch270,26:Pitch180Yaw90,27:Pitch180Yaw270,28:Roll90Pitch90,29:Roll180Pitch90,30:Roll270Pitch90,31:Roll90Pitch180,32:Roll270Pitch180,33:Roll90Pitch270,34:Roll180Pitch270,35:Roll270Pitch270,36:Roll90Pitch180Yaw90,37:Roll90Yaw270,38:Yaw293Pitch68Roll180,39:Pitch315,40:Roll90Pitch315,41:Custom
```
[the remark this answers is not in the document]
Comment (jaxxzer) [reply libraries/AP_AHRS/AP_AHRS.cpp]: gah I thought of this during the fixup then forgot it immediately

File: libraries/AP_InertialSensor/AP_InertialSensor_Backend.cpp [hunk 7]
```diff
@@ -105,7 +105,11 @@ void AP_InertialSensor_Backend::_rotate_and_correct_accel(uint8_t instance, Vect
     accel.z *= accel_scale.z;
 
     // rotate to body frame
-    accel.rotate(_imu._board_orientation);
+    if (_imu._board_orientation == ROTATION_CUSTOM) {
```
[the remark this answers is not in the document]
Comment (jaxxzer) [reply libraries/AP_InertialSensor/AP_InertialSensor_Backend.cpp]: yes

**Comment by jaxxzer** [comment]
Ok, all reviews addressed.

File: libraries/AP_InertialSensor/AP_InertialSensor.cpp [hunk 8]
```diff
@@ -1814,7 +1815,11 @@ bool AP_InertialSensor::get_fixed_mount_accel_cal_sample(uint8_t sample_num, Vec
         return false;
     }
     _accel_calibrator[_acc_body_aligned-1].get_sample_corrected(sample_num, ret);
-    ret.rotate(_board_orientation);
+    if (_board_orientation == ROTATION_CUSTOM && _custom_rotation) {
+      ret = *_custom_rotation * ret;
+    } else {
+      ret.rotate(_board_orientation);
```
Comment (OXINARF) [inline libraries/AP_InertialSensor/AP_InertialSensor.cpp]: Only noticed now, the two _ret_ above are wrongly indented, we use 4 spaces.

File: libraries/AP_InertialSensor/AP_InertialSensor.cpp [hunk 9]
```diff
@@ -1839,7 +1844,11 @@ bool AP_InertialSensor::get_primary_accel_cal_sample_avg(uint8_t sample_num, Vec
     }
     avg /= count;
     ret = avg;
-    ret.rotate(_board_orientation);
+    if (_board_orientation == ROTATION_CUSTOM && _custom_rotation) {
+      ret = *_custom_rotation * ret;
+    } else {
+      ret.rotate(_board_orientation);
```
Comment (OXINARF) [inline libraries/AP_InertialSensor/AP_InertialSensor.cpp]: Same here.

**Comment by jaxxzer** [comment]
Passing again!

**Comment by jaxxzer** [comment]
All reviews addressed here.

**Comment by OXINARF** [comment]
Rebased, conflict fixed, merged. Thanks!

**Comment by jaxxzer** [comment]
Thank you for the help @OXINARF

**Pull request closed without merging**